applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,378 @@
+        Matrix LHS;
+        Vector RHS;
+        Vector disp;
+        Vector external_force;
+        Vector external_force_previous_step;
+
+        Vector disp_previous_step;
+        Vector disp_increment;
+        Vector average_load;
+
+        // TODO Mahmoud: Calculation using the exact value for the external force at the last time step, not just an approximation
+        for (auto& elem_i : rModelPart.Elements())
+        {
+            elem_i.GetValuesVector(disp,0);
+            elem_i.GetValuesVector(disp_previous_step, 1);
+            elem_i.CalculateLocalSystem(LHS, RHS, r_current_process_info);
+
+            disp_increment = disp - disp_previous_step;
+            external_force = -1.0 * RHS;
+            external_force_previous_step = external_force - prod(LHS , disp_increment);
+            average_load = 0.5 * (external_force + external_force_previous_step);
+
+            response_increment_value += inner_prod(average_load , disp_increment);
+        }
+        return response_increment_value;
+        KRATOS_CATCH("");
+    }
+
+    void AdjointNonlinearStrainEnergyResponseFunction::CalculateGradient(const Element& rAdjointElement,
+                                    const Matrix& rResidualGradient,
+                                    Vector& rResponseGradient,
+                                    const ProcessInfo& rProcessInfo)
+    {
+        KRATOS_TRY;
+
+        if (rResponseGradient.size() != rResidualGradient.size1())
+            rResponseGradient.resize(rResidualGradient.size1(), false);
+
+        rResponseGradient.clear();
+
+        KRATOS_CATCH("");
+    }
+
+    // \frac{1}{2} (f_{ext}_i - f_{ext}_i-1) + frac{1}{2} (u^T_i - u^T_i-1) \cdot ( \frac{\partial f_{ext}_i}{\partial u} + frac{\partial f_{ext}_i-1}{\partial u})
+    void AdjointNonlinearStrainEnergyResponseFunction::CalculateGradient(const Condition& rAdjointCondition,
+                                    const Matrix& rResidualGradient,
+                                    Vector& rResponseGradient,
+                                    const ProcessInfo& rProcessInfo)
+    {
+        KRATOS_TRY;
+
+        ProcessInfo process_info = rProcessInfo;
+
+        const SizeType number_of_nodes = rAdjointCondition.GetGeometry().size();
+        const SizeType dimension = rAdjointCondition.GetGeometry().WorkingSpaceDimension();
+        const SizeType vec_size = number_of_nodes * dimension;
+
+        Vector displacement = ZeroVector(vec_size);
+        Vector displacement_previous_step = ZeroVector(vec_size);
+        double delta = rAdjointCondition.GetValue(PERTURBATION_SIZE);
+        Vector RHS;
+
+        // getting the displacements for the current and the previous time steps
+        int i = 0;
+        for (auto& node_i : rAdjointCondition.GetGeometry())
+        {
+            project(displacement, range(i, dimension)) = rAdjointCondition.GetGeometry()[node_i].FastGetSolutionStepValue(DISPLACEMENT);
```

Comment:
I think these operators are not supported by AMatrix, right @jcotela @pooyan-dadvand ?
=> if so then don't use them